packages/material-ui-lab/src/MasonryItem/MasonryItem.js

```diff
@@ -0,0 +1,166 @@
+import { styled, useThemeProps } from '@material-ui/core/styles';
+import ResizeObserver from 'resize-observer-polyfill';
+import { getMasonryItemUtilityClass } from './masonryItemClasses';
+import MasonryContext from '../Masonry/MasonryContext';
+
+const useUtilityClasses = (styleProps) => {
+  const { classes } = styleProps;
+
+  const slots = {
+    root: ['root'],
+  };
+
+  return composeClasses(slots, getMasonryItemUtilityClass, classes);
+};
+
+export const style = ({ styleProps, theme }) => {
+  let styles = {
+    width: '100%',
+    [`& > *`]: {
+      // all contents should have a width of 100%
+      width: '100%',
+    },
+    visibility: styleProps.height ? 'visible' : 'hidden',
+    gridColumnEnd: `span ${styleProps.columnSpan}`,
+  };
+
+  const base = Object.keys(theme.breakpoints.values).reduce((acc, breakpoint) => {
+    if (styleProps.spacing[breakpoint] != null) {
+      acc[breakpoint] = true;
+    }
+    return acc;
+  }, {});
+  const spacingValues = resolveBreakpointValues({ values: styleProps.spacing, base });
+  const transformer = createUnarySpacing(theme);
+  const styleFromPropValue = (propValue) => {
+    const gap = Number(getValue(transformer, propValue).replace('px', ''));
+    const rowSpan = Math.ceil(styleProps.height + gap);
+    return {
+      gridRowEnd: `span ${rowSpan}`,
+      paddingBottom: gap - 1,
+    };
+  };
+  styles = deepmerge(styles, handleBreakpoints({ theme }, spacingValues, styleFromPropValue));
+
+  return styles;
+};
+
+const MasonryItemRoot = styled('div', {
+  name: 'MuiMasonryItem',
+  slot: 'Root',
+  overridesResolver: (props, styles) => {
+    return [styles.root];
+  },
+})(style);
+
+const MasonryItem = React.forwardRef(function MasonryItem(inProps, ref) {
+  const props = useThemeProps({
+    props: inProps,
+    name: 'MuiMasonryItem',
+  });
+
+  const masonryItemRef = React.useRef(null);
+
+  const { spacing = 1, documentReady = false } = React.useContext(MasonryContext);
+  const { children, className, component = 'div', columnSpan = 1, height = 0, ...other } = props;
+  const [styleProps, setStyleProps] = React.useState({
+    ...props,
+    spacing,
+    columnSpan,
+    height,
+  });
+
+  const classes = useUtilityClasses(styleProps);
+
+  const computeHeight = () => {
+    const child = masonryItemRef.current.firstChild;
+    setStyleProps({
+      ...styleProps,
+      height: child?.getBoundingClientRect().height,
+    });
+  };
+  const resizeObserver = React.useRef(new ResizeObserver(computeHeight));
+
+  React.useEffect(() => {
+    if (documentReady) {
+      computeHeight();
+    }
+  }, [documentReady]); // eslint-disable-line
+
+  React.useEffect(() => {
+    if (resizeObserver?.current && masonryItemRef?.current) {
+      const observer = resizeObserver.current;
+      const item = masonryItemRef.current;
+      observer.observe(item);
+      return () => {
+        observer.unobserve(item);
+      };
+    }
+    return true;
+  }, [masonryItemRef]);
```

Comment:
You should get an eslint warning here AFAIK. Refs are stable

```suggestion
  }, []);
```